File: gift_wrapping/jarvis.cpp
/*-------------jarvis.cpp-----------------------------------------------------//
* 
* Purpose: To implement a simple gift-wrapping / convex hull algorithm (jarvis)
*
*   Notes: TEST ANGLE FUNCTION!!!
*          Fix the gift_wrapping syntax
*
*-----------------------------------------------------------------------------*/

#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <cmath>

// Struct
struct pos{
    double x, y;
};

// Struct to hold simulation data
struct parameter{
    std::vector<pos> hull, points;
};

// distance function between points
double dist(pos a, pos b){
    return sqrt((a.x - b.x)*(a.x-b.x)+(a.y-b.y)*(a.y-b.y)) ;
}

// Finding the angle between 3 points
double angle(pos A, pos B, pos C){
    double a = dist(B,C);
    double b = dist(A,C);
    double c = dist(A,B);
    return acos((a*a - b*b - c*c)/(2*a*c));
}

// Function to initialize random points
parameter init(int num); 

// Function to wrap the points with a hull
void gift_wrap(parameter &par);

/*----------------------------------------------------------------------------//
* MAIN
*-----------------------------------------------------------------------------*/

int main(){

    parameter par = init(100);

    for (size_t i = 0; i < par.points.size(); i++){
        std::cout << par.points[i].x << '\t' << par.points[i].y << '\n';
    }
    std::cout << par.hull[0].x << '\t' << par.hull[0].y << '\n';
    //gift_wrap(par);
}

/*----------------------------------------------------------------------------//
* SUBROUTINES
*-----------------------------------------------------------------------------*/

// Function to initialize random points
parameter init(int num){
    parameter par;
    par.points.reserve(num);

    // Creating random device
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_real_distribution<double>
        pos_dist(0,1);

    pos tmp;
    for (int i = 0; i < num; i++){
        tmp.x = pos_dist(gen);
        tmp.y = pos_dist(gen);
        par.points.push_back(tmp);
    }

    // Find the far left point 
    std::sort(par.points.begin(), par.points.end(), 
              [](const pos a, const pos b){return a.x < b.x;});
    return par;
}

// Function to wrap the points with a hull
void gift_wrap(parameter &par){
    // expand the hull
    pos curr_hull, prev_hull, next_hull;
    prev_hull.x = 0.0;
    prev_hull.y = 0.0;
    double threshold = 0.001, angle_tmp, final_angle;
    int i = 0;
    while (dist(curr_hull, par.hull[0]) > threshold){
        if (i == 0){
            curr_hull = par.hull[0];
        }

        final_angle = 2*M_PI;
        for (int j = 0; j < par.points.size(); j++){
             angle_tmp = angle(prev_hull, curr_hull, par.points[j]);
             if (angle_tmp < final_angle){
                 final_angle = angle_tmp;
                 next_hull = par.points[j];
             }
        }
        par.hull.push_back(next_hull);
        prev_hull = curr_hull;
        curr_hull = next_hull;
    }
}

File: gift_wrapping/jarvis.jl
struct Pos
    x::Float64
    y::Float64
end

function jarvis_angle(point1::Pos, point2::Pos, point3::Pos)
    vec1 = Pos(point2.x - point1.x, point2.y - point1.y)
    vec2 = Pos(point3.x - point2.x, point3.y - point2.y)
    mag1 = sqrt(vec1.x*vec1.x + vec1.y*vec1.y)
    mag2 = sqrt(vec2.x*vec2.x + vec2.y*vec2.y)
    ret_angle = acos((vec1.x*vec2.x + vec1.y*vec2.y)/(mag1*mag2))
    return ret_angle
end

function jarvis_march(points::Vector{Pos})
    hull = Vector{Pos}()

    # sorting array based on leftmost point
    sort!(points, by = item -> item.x)
    push!(hull, points[1])

    i = 1
    curr_point = points[2]

    # Find angle between points
    curr_theta = jarvis_angle(Pos(0,0), hull[1], curr_point)
    #while (curr_point != hull[1])
    while (length(hull) < 4)
        println(hull)
        for point in points
                theta = 0.0
            if (i == 1)
                if (hull[i] != point)
                    theta = jarvis_angle(Pos(0,0), hull[i], point)
                end
            else
                if (hull[i] != point && hull[i-1] != point)
                    theta = jarvis_angle(hull[i-1], hull[i], point)
                end
            end
            println(hull[i])
            println(point)
            println(curr_theta)
            println(theta)
            println()
            if (theta > curr_theta)
                curr_point = point
                curr_theta = theta
            end
        end
        push!(hull, curr_point)
        curr_theta = 0
        i += 1
    end

    return hull
end

function main()

    points = [Pos(2,1.5), Pos(1, 1), Pos(2, 4), Pos(3, 1)]
    hull = jarvis_march(points)
    println(hull)
end

println("angle is:")
println(jarvis_angle(Pos(0,0), Pos(1,0), Pos(1,1)))
println("angle is:")
println(jarvis_angle(Pos(0,0), Pos(1,0), Pos(2,1)))

main()
